### PR TITLE
Add Session property to ignore non-readable hive partitions and pass the warning to the user

### PR DESCRIPTION
**Pull request by Naveen007**
Depended by https://github.com/facebookexternal/presto-facebook/pull/1124

Add Session property to ignore non-readable hive partitions and pass that as a warning back to the user

Hive Config
hive.ignore-unreadable-partition

Session property
ignore_unreadable_partition

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by Naveen007** [comment]
#https://github.com/prestodb/presto/issues/14870

File: presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorSplitManager.java [hunk 1]
```diff
@@ -47,10 +51,11 @@ ConnectorSplitSource getSplits(
          * splits without any performance loss.  Non-remotely accessible splits always
          * need to provide host addresses.
          */
-        public SplitSchedulingContext(SplitSchedulingStrategy splitSchedulingStrategy, boolean schedulerUsesHostAddresses)
+        public SplitSchedulingContext(SplitSchedulingStrategy splitSchedulingStrategy, boolean schedulerUsesHostAddresses, List<PrestoWarning> warnings)
```
Comment (highker) [inline presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorSplitManager.java:54]: Just move `WarningCollector` to spi and tunnel it here.

File: presto-raptor/src/test/java/com/facebook/presto/raptor/metadata/TestRaptorSplitManager.java [hunk 2]
```diff
@@ -215,7 +217,8 @@ private void deleteShardNodes()
     private static ConnectorSplitSource getSplits(RaptorSplitManager splitManager, ConnectorTableLayoutResult layout)
     {
         ConnectorTransactionHandle transaction = new RaptorTransactionHandle();
-        return splitManager.getSplits(transaction, SESSION, layout.getTableLayout().getHandle(), new SplitSchedulingContext(UNGROUPED_SCHEDULING, true));
+        List<PrestoWarning> warningList = new ArrayList<PrestoWarning>();
```
Comment (highker) [inline presto-raptor/src/test/java/com/facebook/presto/raptor/metadata/TestRaptorSplitManager.java:220]: We don't usually use list to denote a collector. We use explicit collectors. Objects in our codebase are mostly immutable whenever necessary.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/PlanBuilder.java [hunk 3]
```diff
@@ -302,147 +306,6 @@ public CallExpression comparison(OperatorType operatorType, RowExpression left,
...
-    {
-        private final TypeProvider types;
-        private PlanNode source;
-        private Map<VariableReferenceExpression, Aggregation> assignments = new HashMap<>();
-        private AggregationNode.GroupingSetDescriptor groupingSets;
-        private List<VariableReferenceExpression> preGroupedVariables = new ArrayList<>();
-        private Step step = Step.SINGLE;
-        private Optional<VariableReferenceExpression> hashVariable = Optional.empty();
-        private Optional<VariableReferenceExpression> groupIdVariable = Optional.empty();
-        private Session session = testSessionBuilder().build();
```
Comment (highker) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/PlanBuilder.java:315]: I see a lot of unrelated changes in this PR. We need to leave them untouched. Given there are too many of them, we could either create a new PR or clean up this one. Feel free to pick whichever one is easier. Given the `WarningCollector` move will cause a lot of file changes, also feel free to do it first (either a new PR or a new commit in a PR). Then, we can follow with the actual change.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java [hunk 4]
```diff
@@ -369,7 +381,12 @@ public CounterStat getHighMemorySplitSource()
                     // verify partition is not marked as non-readable
                     String partitionNotReadable = partition.getParameters().get(OBJECT_NOT_READABLE);
                     if (!isNullOrEmpty(partitionNotReadable)) {
-                        throw new HiveNotReadableException(tableName, Optional.of(partName), partitionNotReadable);
+                        if (!shouldIgnoreUnreadablePartition(session)) {
+                            throw new HiveNotReadableException(tableName, Optional.of(partName), partitionNotReadable);
+                        }
+                        warningCollector.add(new PrestoWarning(HiveWarningCode.PARTITION_NOT_READABLE,
```
Comment (highker) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java:387]: - move the first param to it's own line
- static import `PARTITION_NOT_READABLE`
- Can we add a unit test for this path?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java [hunk 5]
```diff
@@ -369,7 +381,12 @@ public CounterStat getHighMemorySplitSource()
...
+                                format("Table '%s' partition '%s' is not readable: %s", tableName, Optional.of(partName).get(), partitionNotReadable)));
```
Comment (highker) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java:388]: just `partName`. Optional.of(partName).get() === partName

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientFileMetastore.java [hunk 6]
```diff
@@ -13,17 +13,56 @@
...
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static org.apache.hadoop.hive.common.FileUtils.makePartName;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertNotNull;
+import static org.testng.Assert.fail;
 
 public class TestHiveClientFileMetastore
         extends AbstractTestHiveClientLocal
 {
+    private final String tempTable = "tempTable";
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientFileMetastore.java:61]: inline this const

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientFileMetastore.java [hunk 7]
```diff
@@ -13,17 +13,56 @@
...
+    public static final ConnectorSplitManager.SplitSchedulingContext SPLIT_SCHEDULING_CONTEXT = new ConnectorSplitManager.SplitSchedulingContext(UNGROUPED_SCHEDULING,
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientFileMetastore.java:62]: - One param per line and leave the first line empty
- static import SplitSchedulingContext

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientFileMetastore.java [hunk 8]
```diff
@@ -66,4 +105,117 @@ public void testTransactionDeleteInsert()
...
+
+                ColumnHandle dsColumn = metadata.getColumnHandles(session, tableHandle).get("ds");
+                assertNotNull(dsColumn);
+
+                ConnectorTableLayout tableLayout = getTableLayout(session, metadata, tableHandle, Constraint.alwaysTrue(), transaction);
+                splitManager.getSplits(transaction.getTransactionHandle(), session, tableLayout.getHandle(), SPLIT_SCHEDULING_CONTEXT);
+                assertNotNull(SPLIT_SCHEDULING_CONTEXT.getWarningCollector());
+                assertEquals(SPLIT_SCHEDULING_CONTEXT.getWarningCollector().getWarnings().size(), 1);
+            }
+        }
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientFileMetastore.java:156]: add a catch here and fail the test

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientFileMetastore.java [hunk 9]
```diff
@@ -66,4 +105,117 @@ public void testTransactionDeleteInsert()
...
         // FileHiveMetastore has various incompatibilities
     }
+
+    @Test
+    public void testPartitionNotReadable()
+            throws Exception
+    {
+        SchemaTableName tableName = temporaryTable(tempTable);
+        ImmutableMap dynamicPartitionParameters = ImmutableMap.of(
+                HiveSplitManager.OBJECT_NOT_READABLE, "Testing Unreadable Partition");
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientFileMetastore.java:115]: - static import OBJECT_NOT_READABLE
- move this together with the previous line

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientFileMetastore.java [hunk 10]
```diff
@@ -66,4 +105,117 @@ public void testTransactionDeleteInsert()
...
+                assertNotNull(SPLIT_SCHEDULING_CONTEXT.getWarningCollector());
+                assertEquals(SPLIT_SCHEDULING_CONTEXT.getWarningCollector().getWarnings().size(), 1);
+            }
+        }
+        finally {
+            dropTable(tableName);
+        }
+    }
+
+    protected void createDummyPartitionedTable(SchemaTableName tableName, List<ColumnMetadata> columns, ImmutableMap dynamicPartitionParameters)
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientFileMetastore.java:162]: private; same for the ones below.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientFileMetastore.java [hunk 11]
```diff
@@ -66,4 +105,117 @@ public void testTransactionDeleteInsert()
...
+
+        List<PartitionWithStatistics> partitions = ImmutableList.of(firstPartitionName, secondPartitionName)
+                .stream()
+                .map(partitionName -> new PartitionWithStatistics(createDummyPartition(table, partitionName), partitionName, PartitionStatistics.empty()))
+                .collect(toImmutableList());
+        metastoreClient.addPartitions(tableName.getSchemaName(), tableName.getTableName(), partitions);
+        metastoreClient.updatePartitionStatistics(tableName.getSchemaName(), tableName.getTableName(), firstPartitionName, currentStatistics -> EMPTY_TABLE_STATISTICS);
+        metastoreClient.updatePartitionStatistics(tableName.getSchemaName(), tableName.getTableName(), secondPartitionName, currentStatistics -> EMPTY_TABLE_STATISTICS);
+
+        List<PartitionWithStatistics> partitionsNotReable = ImmutableList.of(thirdPartitionName)
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientFileMetastore.java:187]: typo Reable

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientFileMetastore.java [hunk 12]
```diff
@@ -66,4 +105,117 @@ public void testTransactionDeleteInsert()
...
+    }
+
+    protected Partition createDummyPartition(Table table, String partitionName, ImmutableMap dynamicPartitionParameters)
+    {
+        return createDummyPartition(table, partitionName, Optional.empty(), dynamicPartitionParameters);
+    }
+
+    protected Partition createDummyPartition(Table table, String partitionName, Optional<HiveBucketProperty> bucketProperty, ImmutableMap dynamicPartitionParameters)
+    {
+        ImmutableMap staticPartitionParameters = ImmutableMap.of(
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientFileMetastore.java:202]: `Map<..., ..>`

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientFileMetastore.java [hunk 13]
```diff
@@ -66,4 +106,119 @@ public void testTransactionDeleteInsert()
     {
         // FileHiveMetastore has various incompatibilities
     }
+
+    @Test
+    public void testPartitionNotReadable()
+            throws Exception
+    {
+        SchemaTableName tableName = temporaryTable("tempTable");
+        ImmutableMap<String, String> dynamicPartitionParameters = ImmutableMap.of(OBJECT_NOT_READABLE, "Testing Unreadable Partition");
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientFileMetastore.java:115]: `Map<String, String>`

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientFileMetastore.java [hunk 14]
```diff
@@ -13,17 +13,57 @@
...
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static org.apache.hadoop.hive.common.FileUtils.makePartName;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertNotNull;
+import static org.testng.Assert.fail;
 
 public class TestHiveClientFileMetastore
         extends AbstractTestHiveClientLocal
 {
+    public static final SplitSchedulingContext SPLIT_SCHEDULING_CONTEXT = new SplitSchedulingContext(
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientFileMetastore.java:62]: private

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientFileMetastore.java [hunk 15]
```diff
@@ -66,4 +106,119 @@ public void testTransactionDeleteInsert()
     {
         // FileHiveMetastore has various incompatibilities
     }
+
+    @Test
+    public void testPartitionNotReadable()
+            throws Exception
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientFileMetastore.java:112]: this is redundant

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientFileMetastore.java [hunk 16]
```diff
@@ -66,4 +106,119 @@ public void testTransactionDeleteInsert()
...
+        }
+        catch (Exception e) {
+            fail("Exception not expected");
+        }
+        finally {
+            dropTable(tableName);
+        }
+    }
+
+    private void createDummyPartitionedTable(SchemaTableName tableName, List<ColumnMetadata> columns, ImmutableMap dynamicPartitionParameters)
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientFileMetastore.java:165]: `ImmutableMap` -> `Map<String, String>`

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientFileMetastore.java [hunk 17]
```diff
@@ -66,4 +106,119 @@ public void testTransactionDeleteInsert()
...
+
+        List<PartitionWithStatistics> partitionsNotReadable = ImmutableList.of(thirdPartitionName)
+                .stream()
+                .map(partitionName -> new PartitionWithStatistics(createDummyPartition(table, partitionName, dynamicPartitionParameters), partitionName, PartitionStatistics.empty()))
+                .collect(toImmutableList());
+        metastoreClient.addPartitions(tableName.getSchemaName(), tableName.getTableName(), partitionsNotReadable);
+        metastoreClient.updatePartitionStatistics(tableName.getSchemaName(), tableName.getTableName(), thirdPartitionName, currentStatistics -> EMPTY_TABLE_STATISTICS);
+    }
+
+    private Partition createDummyPartition(Table table, String partitionName, ImmutableMap dynamicPartitionParameters)
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientFileMetastore.java:198]: ImmutableMap -> Map<String, String>

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientFileMetastore.java [hunk 18]
```diff
@@ -66,4 +106,119 @@ public void testTransactionDeleteInsert()
...
+    {
+        return createDummyPartition(table, partitionName, Optional.empty(), dynamicPartitionParameters);
+    }
+
+    private Partition createDummyPartition(Table table, String partitionName, Optional<HiveBucketProperty> bucketProperty, ImmutableMap dynamicPartitionParameters)
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientFileMetastore.java:203]: ImmutableMap -> Map<String, String>

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientFileMetastore.java [hunk 19]
```diff
@@ -66,4 +106,119 @@ public void testTransactionDeleteInsert()
...
+    {
+        ImmutableMap<String, String> staticPartitionParameters = ImmutableMap.of(
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientFileMetastore.java:205]: Map<String, String>

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientFileMetastore.java [hunk 20]
```diff
@@ -66,4 +106,119 @@ public void testTransactionDeleteInsert()
...
+                PRESTO_VERSION_NAME, "testversion",
+                PRESTO_QUERY_ID_NAME, "20200101_123456_00001_x1y2z");
+        ImmutableMap<String, String> partitionParameters = ImmutableMap.<String, String>builder()
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientFileMetastore.java:208]: Map<String, String>

**Pull request merged**